### PR TITLE
stm32mp1: PWR regulators thru SCMI

### PR DESCRIPTION
**Pull request by etienne-lms**
Supersedes  #4131.

### CODE REVIEW

File: core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr.c [hunk 1]
```diff
@@ -3,13 +3,86 @@
...
+
+#define PWR_CR3_USB33_EN	BIT(24)
+#define PWR_CR3_USB33_RDY	BIT(26)
+#define PWR_CR3_REG18_EN	BIT(28)
+#define PWR_CR3_REG18_RDY	BIT(29)
+#define PWR_CR3_REG11_EN	BIT(30)
+#define PWR_CR3_REG11_RDY	BIT(31)
+
+struct pwr_regu_desc {
+	unsigned long level_mv;
```
Comment (jenswi-linaro) [inline core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr.c:21]: I'm curious about the choice of type. Why `unsigned long` and not `unsigned int` or `uint32_t`?

File: core/arch/arm/plat-stm32mp1/scmi_server.c [hunk 2]
```diff
@@ -448,6 +492,195 @@ int32_t plat_scmi_rd_set_state(unsigned int agent_id, unsigned int scmi_id,
...
+ * Platform SCMI voltage domains
+ */
+static struct stm32_scmi_voltd *find_voltd(unsigned int agent_id,
+					   unsigned int scmi_id)
+{
+	const struct scmi_agent_resources *resource = find_resource(agent_id);
+	size_t n = 0;
+
+	if (resource) {
+		for (n = 0U; n < resource->voltd_count; n++)
```
Comment (jenswi-linaro) [inline core/arch/arm/plat-stm32mp1/scmi_server.c:505]: A plain `0` should be good enough here.

File: core/arch/arm/plat-stm32mp1/scmi_server.c [hunk 3]
```diff
@@ -448,6 +492,195 @@ int32_t plat_scmi_rd_set_state(unsigned int agent_id, unsigned int scmi_id,
...
+	/* Currently non-secure is allowed to access all PWR regulators */
+	if (!voltd)
+		return NULL;
+
+	return voltd->name;
+}
+
+static enum pwr_regulator pwr_scmi_to_regu_id(struct stm32_scmi_voltd *voltd)
+{
+	size_t id = voltd->priv_id[0] - '0';
```
Comment (jenswi-linaro) [inline core/arch/arm/plat-stm32mp1/scmi_server.c:537]: You have a hidden dependency to the private string name of the voltage domain. Sounds difficult to maintain.

File: core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr.c (hunk 1 above)
Comment (etienne-lms) [reply core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr.c:21]: No strong constraint here. `unsigned int` would fit.
I'll fix as it looks more... simple.

File: core/arch/arm/plat-stm32mp1/scmi_server.c (hunk 2 above)
Comment (etienne-lms) [reply core/arch/arm/plat-stm32mp1/scmi_server.c:505]: :)  right!

File: core/arch/arm/plat-stm32mp1/scmi_server.c (hunk 3 above)
Comment (etienne-lms) [reply core/arch/arm/plat-stm32mp1/scmi_server.c:537]: Ok, i'll make it explicit.

**Comment by jenswi-linaro** [comment]
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

File: core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr.h [hunk 4]
```diff
@@ -17,6 +18,17 @@
...
+	PWR_REG18,
+	PWR_USB33,
+	PWR_REGU_COUNT
+};
+
 vaddr_t stm32_pwr_base(void);
 
+unsigned int stm32mp1_pwr_regulator_mv(enum pwr_regulator id);
+void stm32mp1_pwr_regulator_set_state(enum pwr_regulator id, bool enable);
+bool stm32mp1_pwr_regulator_is_enable(enum pwr_regulator id);
```
Comment (jforissier) [inline core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr.h:32]: enable**d**?

Comment (etienne-lms) [reply core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr.h:32]: Right, thanks.

**Comment by etienne-lms** [comment]
Comment addressed. I'll squash the fixup commits.

**Comment by etienne-lms** [comment]
Squashed and fixed coding style issues (misplaced space chars).
@jenswi-linaro tag applied.

**Comment by jforissier** [comment]
@etienne-lms do you expect this in 3.11.0? It is not a bug fix or a trivial change, so in principle it would not qualify for merging after -rc1. On the other hand it can only impact stm32mp1 and I don't want to limit the usefulness of the release by being too strict.

**Comment by etienne-lms** [comment]
Yes, I would be pleased to see these regulators in 3.11.0 tag.

**Comment by jforissier** [comment]
OK

**Pull request merged**